### ISSUE TITLE
GraalVM Check Intermittently Fails for Pull Requests in Module-Ballerina-Persist

### ISSUE THREAD

**Issue by sahanHe** [ballerina-platform/ballerina-standard-library#4451]
**Description:**
GraalVM checks are intermittently failing with the following exit code for pull requests and `Check GraalVM` workflow with the following log.

```
========================================================================================================================
GraalVM Native Image: Generating 'persist' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                   (18.1s @ 0.47GB)
 Version info: 'GraalVM 22.3.2 Java 11 CE'
 Java version info: '11.0.19+7-jvmci-22.3-b18'
 C compiler: gcc (linux, x86_64, 11.3.0)
 Garbage collector: Serial GC
 2 user-specific feature(s)
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - io.ballerina.stdlib.crypto.svm.BouncyCastleFeature
[2/7] Performing analysis...  [*************]                                                          (287.8s @ 4.26GB)
  35,158 (95.41%) of 36,850 classes reachable
 143,509 (84.38%) of 170,082 fields reachable
 204,781 (78.61%) of 260,491 methods reachable
   1,532 classes,   778 fields, and 3,686 methods registered for reflection
      92 classes,    97 fields, and    69 methods registered for JNI access
       7 native libraries: dl, m, pthread, rt, stdc++, z
[3/7] Building universe...                                                                              (49.0s @ 4.62GB)
[4/7] Parsing methods...      [*******]                                                                 (57.6s @ 2.98GB)
[5/7] Inlining methods...     [***]                                                                     (19.0s @ 3.49GB)
[6/7] Compiling methods...    [************************]                                               (617.2s @ 4.43GB)
Error: The operation was canceled.
```
source - [https://github.com/ballerina-platform/module-ballerina-persist/actions/runs/4977461219/jobs/8907480877#step:9:185](https://github.com/ballerina-platform/module-ballerina-persist/actions/runs/4977461219/jobs/8907480877#step:9:185)

At times of success, it gives the following log

```
========================================================================================================================
GraalVM Native Image: Generating 'persist' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                   (22.9s @ 0.49GB)
 Version info: 'GraalVM 22.3.2 Java 11 CE'
 Java version info: '11.0.19+7-jvmci-22.3-b18'
 C compiler: gcc (linux, x86_64, 11.3.0)
 Garbage collector: Serial GC
 2 user-specific feature(s)
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - io.ballerina.stdlib.crypto.svm.BouncyCastleFeature
[2/7] Performing analysis...  [*************]                                                          (389.8s @ 3.59GB)
  35,158 (95.41%) of 36,850 classes reachable
 143,509 (84.38%) of 170,082 fields reachable
 [204](https://github.com/sahanHe/module-ballerina-persist/actions/runs/4977782568/jobs/8907241714#step:9:205),781 (78.61%) of 260,491 methods reachable
   1,532 classes,   778 fields, and 3,686 methods registered for reflection
      92 classes,    97 fields, and    69 methods registered for JNI access
       7 native libraries: dl, m, pthread, rt, stdc++, z
[3/7] Building universe...                                                                              (68.1s @ 4.57GB)
[4/7] Parsing methods...      [*********]                                                               (87.0s @ 3.57GB)
[5/7] Inlining methods...     [***]                                                                     (27.3s @ 3.08GB)
[6/7] Compiling methods...    [*******************************]                                        (957.4s @ 4.45GB)
GC warning: 514.9s spent in 107 GCs during the last stage, taking up 53.78% of the time.
            Please ensure more than 6.16GB of memory is available for Native Image
            to reduce GC overhead and improve image build time.
[7/7] Creating image...                                                                                 (83.3s @ 5.45GB)
 174.38MB (64.86%) for code area:   149,385 compilation units
  85.18MB (31.68%) for image heap:  625,079 objects and 100 resources
   9.31MB ( 3.46%) for other data
 268.88MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
  21.69MB ballerina.persist$test/1.tests                      28.32MB byte[] for code metadata
  17.32MB ballerina.http/2                                    13.28MB byte[] for embedded resources
  12.77MB ballerina.persist/1                                  9.58MB java.lang.Class
  11.76MB ballerinax.googleapis$0046sheets/3                   7.47MB byte[] for java.lang.String
   8.96MB ballerina.persist$test/1                             6.21MB java.lang.String
   7.47MB ballerinai.transaction/0                             4.48MB byte[] for general heap data
   4.49MB ballerina.http/2.types                               3.22MB com.oracle.svm.core.hub.DynamicHubCompanion
   3.42MB ballerina.test/0                                     1.83MB byte[] for reflection metadata
   2.77MB ballerina.io/1                                       1.30MB java.lang.String[]
   2.46MB ballerina.sql/1                                      1.29MB c.o.svm.core.hub.DynamicHub$ReflectionMetadata
  80.23MB for 1018 more packages                               7.52MB for 5[220](https://github.com/sahanHe/module-ballerina-persist/actions/runs/4977782568/jobs/8907241714#step:9:221) more object types
------------------------------------------------------------------------------------------------------------------------
                      675.5s (38.9% of total time) in [223](https://github.com/sahanHe/module-ballerina-persist/actions/runs/4977782568/jobs/8907241714#step:9:224) GCs | Peak RSS: 6.16GB | CPU load: 1.13
------------------------------------------------------------------------------------------------------------------------
```

Have checked increasing the heap memory allocation but still fails intermittently with the given error.
source - [https://github.com/sahanHe/module-ballerina-persist/actions/runs/4977782568/jobs/8907241714](https://github.com/sahanHe/module-ballerina-persist/actions/runs/4977782568/jobs/8907241714)

**Comment by Akshay-De-Silva** [comment]
The build passes on an ubuntu local machine when default_native.properties file is created with the following config:
`NativeImageArgs = -J-Xmx7G`
```
GraalVM Native Image: Generating 'persist' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                   (13.3s @ 0.41GB)
 Version info: 'GraalVM 22.3.2 Java 11 CE'
 Java version info: '11.0.19+7-jvmci-22.3-b18'
 C compiler: gcc (linux, x86_64, 9.4.0)
 Garbage collector: Serial GC
 2 user-specific feature(s)
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - io.ballerina.stdlib.crypto.svm.BouncyCastleFeature
[2/7] Performing analysis...  [*************]                                                          (158.5s @ 3.34GB)
  35,257 (95.42%) of 36,949 classes reachable
 143,579 (84.34%) of 170,238 fields reachable
 204,849 (78.58%) of 260,683 methods reachable
   1,533 classes,   803 fields, and 3,687 methods registered for reflection
      92 classes,    97 fields, and    69 methods registered for JNI access
       7 native libraries: dl, m, pthread, rt, stdc++, z
[3/7] Building universe...                                                                              (14.9s @ 4.42GB)
[4/7] Parsing methods...      [****]                                                                    (20.8s @ 4.99GB)
[5/7] Inlining methods...     [***]                                                                      (8.5s @ 5.25GB)
[6/7] Compiling methods...    [******************]                                                     (336.6s @ 4.87GB)
GC warning: 187.4s spent in 137 GCs during the last stage, taking up 55.67% of the time.
            Please ensure more than 7.94GB of memory is available for Native Image
            to reduce GC overhead and improve image build time.
[7/7] Creating image...                                                                                 (10.5s @ 4.18GB)
 176.66MB (65.09%) for code area:   149,472 compilation units
  85.42MB (31.47%) for image heap:  617,212 objects and 100 resources
   9.32MB ( 3.43%) for other data
 271.40MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
  21.69MB ballerina.persist$test/1.tests                      28.58MB byte[] for code metadata
  17.32MB ballerina.http/2                                    13.28MB byte[] for embedded resources
  12.77MB ballerina.persist/1                                  9.60MB java.lang.Class
  11.76MB ballerinax.googleapis$0046sheets/3                   7.44MB byte[] for java.lang.String
   9.76MB ballerinai.transaction/0                             6.12MB java.lang.String
   8.96MB ballerina.persist$test/1                             4.45MB byte[] for general heap data
   4.49MB ballerina.http/2.types                               3.23MB com.oracle.svm.core.hub.DynamicHubCompanion
   3.42MB ballerina.test/0                                     1.83MB byte[] for reflection metadata
   2.77MB ballerina.io/1                                       1.29MB c.o.svm.core.hub.DynamicHub$ReflectionMetadata
   2.46MB ballerina.sql/1                                      1.26MB java.lang.String[]
  80.22MB for 1018 more packages                               7.39MB for 5343 more object types
------------------------------------------------------------------------------------------------------------------------
                      212.5s (36.6% of total time) in 263 GCs | Peak RSS: 8.12GB | CPU load: 5.75
```

**Comment by TharmiganK** [comment]
Might be an OOM issue in the github runner which caused the github runner to terminate the process. Can we merge this changes to the main branch and check the daily build?

**Comment by TharmiganK** [comment]
Lang has done some improvements to the native build memory usage with transaction in this PR: https://github.com/ballerina-platform/ballerina-lang/pull/40350

This is now merged with `ballerina-lang` `master` branch. And I ran the `GraalVM Check` multiple times in the repo with this lang improvement seems like this intermittent issue is now gone.

@sahanHe Can we check the builds in the PR checks and close this issue?
